File: client/src/model/resume.ts
export interface ResumeView {
    id: string;
    title: string
    created: Date
}

export interface Resume {
    [key: string]: string | undefined | Contact | Date | Array<Employment | Education | Certification | Language>
    id?: string
    title: string
    imageId?: string
    created: Date
    name: string
    position: string;
    contact: Contact;
    skills?: string;
    experiences: Array<Employment>;
    educations: Array<Education>;
    certifications: Array<Certification>;
    summary?: string;
    languages: Array<Language>;
}

export interface Contact {
    [key: string]: string | undefined;
    email: string;
    location?: string;
    phone?: string;
}

export interface Employment {
    [key: string]: string | Date | undefined;
    companyName: string;
    position: string;
    city?: string;
    startDate?: Date;
    endDate?: Date;
    description?: string;
}

export interface Education {
    [key: string]: string | Date | undefined;
    universityName: string,
    degree: string,
    city?: string;
    startDate?: Date;
    endDate?: Date;
    description?: string;
}

export interface Certification {
    [key: string]: string | Date | undefined;
    institution: string;
    name: string;
    link?: string;
    completionDate?: Date;
}

export interface Language {
    [key: string]: string | undefined;
    languageName: string;
    level?: string;
}

export const getEmploymentDefault = (): Employment => ({
    companyName: "",
    position: "",
    city: "",
});

export const getEducationDefault = (): Education => ({
    universityName: "",
    degree: "",
    city: "",
});

export const getCertificationDefault = (): Certification => ({
    name: "",
    institution: "",
    link: "",
});

export const getLanguageDefault = (): Language => ({
    languageName: "",
    level: ""
});

export const resumeDefault: Resume = {
    name: "",
    position: "",
    title: "",
    contact: {
        location: "",
        phone: "",
        email: ""
    },
    created: new Date(),
    experiences: [],
    educations: [],
    certifications: [],
    languages: [],
}

File: api/src/main/kotlin/com/cvmaker/api/model/Resume.kt
package com.cvmaker.api.model

import jakarta.validation.Valid
import jakarta.validation.constraints.NotBlank
import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.mapping.Document
import java.time.LocalDate

@Document("Resumes")
data class Resume(
    @Id
    var id: String?,

    var userEmail: String?,

    @field:NotBlank(message = "title must not be blank")
    val title: String,

    @field:NotBlank(message = "name must not be blank")
    val name: String,

    @field:NotBlank(message = "position must not be blank")
    val position: String,

    val created: LocalDate,

    val contact: Contact,

    var imageId: String?,

    val skills: String?,

    val summary: String?,

    @field:Valid
    val experiences: List<Employment>,

    @field:Valid
    val educations: List<Education>,

    @field:Valid
    val certifications: List<Certification>,

    @field:Valid
    val languages: List<Language>,
)
